File: Python_Development/word_count.py
import sys
import os

#get all words from file
def get_words(file_name):
	words = []
	with open(file_name, "r") as file:
		for line in file:
			line = line.replace(".,;()", " ").lower()
			words += line.split()
	return words

#print length ascending (part A)
def print_length_ascending(words):
	lengths = {len(word) for word in words}
	for length in lengths:
		length_words = [word for word in words if len(word) == length]
		frequency = len(length_words)
		print("Count[{0:0=2d}]={1:0=2d};".format(length, frequency))

#print freqeuncy descending (part B)
def print_frequency_descending(words):
	lengths = {len(word) for word in words}
	temp_frequency = 0
	for length in lengths:
		length_words = [word for word in words if len(word) == length]
		frequency = len(length_words)
		if(temp_frequency < frequency):
			temp_frequency = frequency
	for a in range(temp_frequency, 0, -1):
		for length in lengths:
			length_words = [word for word in words if len(word) == length]
			frequency = len(length_words)
			if(a == frequency):
				print("Count[{0:0=2d}]={1:0=2d};".format(length, frequency))

#print frequency descending with words (part C)
def print_frequency_descending_words(words):
	lengths = {len(word) for word in words}
	temp_frequency = 0
	for length in lengths:
		length_words = [word for word in words if len(word) == length]
		frequency = len(length_words)
		if(temp_frequency < frequency):
			temp_frequency = frequency
	for a in range(temp_frequency, 0, -1):
		for length in lengths:
			length_words = [word for word in words if len(word) == length]
			frequency = len(length_words)
			unique = list(set(length_words))
			unique.sort()
			if(a == frequency):
				sys.stdout.write("Count[{0:0=2d}]={1:0=2d}; (words: ".format(length, frequency))
				for a in range(len(unique)):
					sys.stdout.write("\"{0}\"".format(unique[a]))
					if(a+2 == len(unique)):
						sys.stdout.write(" and ")
					elif(a+1 != len(unique)):
						sys.stdout.write(", ")
				print(")")

#print length ascending with words (part C special case)
def print_length_ascending_words(words):
	lengths = {len(word) for word in words}
	for length in lengths:
		length_words = [word for word in words if len(word) == length]
		frequency = len(length_words)
		unique = list(set(length_words))
		unique.sort()
		sys.stdout.write("Count[{0:0=2d}]={1:0=2d}; (words: ".format(length, frequency))
		for a in range(len(unique)):
			sys.stdout.write("\"{0}\"".format(unique[a]))
			if(a+2 == len(unique)):
				sys.stdout.write(" and ")
			elif(a+1 != len(unique)):
				sys.stdout.write(", ")
		print(")")

def main():
	sort = 0
	print_words = 0
	sort_print_words = 0
	if(len(sys.argv) < 3):
		print("Failure: not enough inptus")
		return -1
	for a in range(len(sys.argv)):
		if(sys.argv[a] == "--infile"):
			file_name = sys.argv[a+1]
			if(not os.path.exists(file_name)):
				print("Failure: could not access file")
				return -1
		if(sys.argv[a] == "--sort"):
			sort = 1
		if(sys.argv[a] == "--print-words"):
			print_words = 1
		sort_print_words = sort + print_words
		if(sort_print_words == 2):
			sort = 0
			print_words = 0
	words = get_words(file_name)
	if(sort_print_words == 0):#part A
		print_length_ascending(words)
	if(sort == 1):#part B
		print_frequency_descending(words)
	if(sort_print_words == 2):#part C
		print_frequency_descending_words(words)
	if(print_words == 1):#part C special case
		print_length_ascending_words(words)
		
main()
#run with python3 word_count.py --infile <input_file> <--sort> <--print-words>

File: Dynamic_Memory/word_count.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>

//create a node for linked list 
struct node{
	char* word;
	int length;
	int frequency;
	struct node* next;
	struct node* down;
};

//free memory allocated for list
void Free_List(struct node** head){
	struct node* current = *head;
	struct node* current_next;
	struct node* temp_down;
	struct node* temp_down_next;
	while(current != NULL){
		current_next = current->next;
		temp_down = current;
		while(temp_down != NULL){
			temp_down_next = temp_down->down;
			free(temp_down->word);
			free(temp_down);
			temp_down = temp_down_next;
		}
		current = current_next;
	}
}

//insert a new node into the linked list
void Insert_Node(struct node** head, struct node* new){
	struct node* current = *head;
	if(current == NULL){//empty list
		*head = new;
		return;
	}
	if(new->length < current->length){//new word length is less than head
			new->next = *head;
			*head = new;
			return;
	}
	while(current != NULL){	
		if(new->length == current->length){//new word is same length as head
			current->frequency += 1;
			if(strcasecmp(new->word, current->word) != 0){
				while(current->down != NULL){
					current = current->down;
					if(strcasecmp(new->word, current->word) == 0){
						return;
					}
				}
				current->down = new;
			}
			return;
		}
		current = current->next;
	}
	current = *head;
	while(current->next != NULL){
		if(new->length < current->next->length){//new word length is less than current's next
			new->next = current->next;
			current->next = new;
			return;
		}
		current = current->next;
	}
	current->next = new;//new word length is greater than all lengths
}

//create a new node to be added to linked list
void Create_Node(struct node** head, char* word){
	struct node* new = malloc(sizeof(*new));
	int length = strlen(word) + 1;
	new->word = (char*)malloc(sizeof(*word));
	strncpy(new->word, word, length);
	new->length = length - 1;
	new->frequency = 1;
	new->next = NULL;
	new->down = NULL;
	Insert_Node(head, new);
}

//get all data from the file (the word, its length, and the frequency)
void Get_Words(struct node** head, char word_buffer[]){
	char* current = strtok(word_buffer, " .,;()\b\t\n"); 
	if(current){
		Create_Node(head, current);
	}
	while(current){
		current = strtok(NULL, " .,;()\b\t\n"); 		
		if(current){
			Create_Node(head, current);
		}
	}
}

//print length ascending (part A)
void Print_Length_Ascending(struct node* head){
	struct node* current = head;
	while(current != NULL){ 
		printf("Count[%02d]=%02d;\n", current->length, current->frequency);
		current = current->next;
	}
}

//print frequency descending (part B)
void Print_Frequency_Descending(struct node* head){
	struct node* current = head;
	struct node* temp = head;
	int frequency = current->frequency;
	while(temp != NULL){
		if(frequency < temp->frequency){
			frequency = temp->frequency;
		}
		temp = temp->next;
	}
	while(frequency > 0){
		current = head;
		while(current != NULL){
			if(frequency == current->frequency){
				printf("Count[%02d]=%02d;\n", current->length, current->frequency);
			}
			current = current->next;
		}
		frequency--;
	}
}

//print frequency descending with words (part C)
void Print_Frequency_Descending_Words(struct node* head){
	struct node* current = head;
	struct node* current_down = head;
	struct node* temp = head;
	int frequency = current->frequency;
	while(temp != NULL){
		if(frequency < temp->frequency){
			frequency = temp->frequency;
		}
		temp = temp->next;
	}
	while(frequency > 0){
		current = head;
		while(current != NULL){ 
			current_down = current;
			if(frequency == current->frequency){	
				printf("Count[%02d]=%02d; ", current->length, current->frequency);
				printf("(words: \"%s\"", current->word);
				while(current_down->down != NULL){
					current_down = current_down->down;
					if(current_down->down != NULL){
						printf(", ");
					}
					if(current_down->down == NULL){
						printf(" and ");
					}
					printf("\"%s\"", current_down->word);
				}
				printf(")\n");
			}
			current = current->next;
		}
		frequency--;
	}
}

//print length ascending with words (part C special case)
void Print_Length_Ascending_Words(struct node* head){
	struct node* current = head;
	struct node* current_down = current;
	while(current != NULL){ 
		current_down = current;
		printf("Count[%02d]=%02d; ", current->length, current->frequency);
		printf("(words: \"%s\"", current->word);
		while(current_down->down != NULL){
			current_down = current_down->down;
			if(current_down->down != NULL){
				printf(",");
			}
			if(current_down->down == NULL){
				printf(" and");
			}
			printf(" \"%s\"", current_down->word);
		}
		printf(")\n");
		current = current->next;
	}
}

int main(int argc, char *argv[]){
	struct node* head = NULL;
	FILE* file_name = NULL;
	int sort;
	int print_words;
	int sort_print_words;
	if(argc < 3){
		printf("Failure: not enough inputs\n");
		return -1; 
	}
	else{
		for(int a = 0; a < argc; a++){//find user inputs
			if(!strcmp(argv[a], "--infile")){
				file_name = fopen(argv[a+1], "r");
				if(file_name == NULL){//could not find file
					printf("Failure: could not access file\n");
					return -1;
				}
			}
			if(!strcmp(argv[a], "--sort")){
				sort = 1;
			}
			if(!strcmp(argv[a], "--print-words")){
				print_words = 1;
			}
			sort_print_words = sort + print_words;
			if(sort_print_words == 2){
				sort = 0;
				print_words = 0;
			}
		}
		fseek(file_name, 0, SEEK_END);
		char word_buffer[ftell(file_name)+1];
		fseek(file_name, 0, SEEK_SET);	
		while(fgets(word_buffer, sizeof(word_buffer), file_name)){
			if(strlen(word_buffer) > 0){
				Get_Words(&head, word_buffer);
			}
		}
		if(sort_print_words == 0){//part A
			Print_Length_Ascending(head);
		}
		else if(sort == 1){//part B
			Print_Frequency_Descending(head);
		}
		else if(sort_print_words == 2){//part C
			Print_Frequency_Descending_Words(head);
		}
		else if(print_words == 1){//part C special case
			Print_Length_Ascending_Words(head);
		}
	}
	fclose(file_name);
	Free_List(&head);
	return 0;
}
//compile with gcc word_count.c
//run with ./a.out --infile <input_file> <--sort> <--print-words>